File: cpp/cluster_management/src/CreateSingleRegion.cpp
#include <aws/core/Aws.h>
#include <aws/core/utils/Outcome.h>
#include <aws/dsql/DSQLClient.h>
#include <aws/dsql/model/CreateClusterRequest.h>
#include <aws/dsql/model/GetClusterRequest.h>
#include <iostream>
#include <thread>
#include <chrono>

using namespace Aws;
using namespace Aws::DSQL;
using namespace Aws::DSQL::Model;

/**
 * Creates a single-region cluster in Amazon Aurora DSQL
 */
CreateClusterResult CreateCluster(const Aws::String& region) {
    // Create client for the specified region
    DSQL::DSQLClientConfiguration clientConfig;
    clientConfig.region = region;
    DSQL::DSQLClient client(clientConfig);
    
    // Create the cluster
    CreateClusterRequest createClusterRequest;
    createClusterRequest.SetDeletionProtectionEnabled(true);
    
    // Add tags
    Aws::Map<Aws::String, Aws::String> tags;
    tags["Name"] = "cpp single region cluster";
    const char* repo = std::getenv("GITHUB_REPOSITORY");
    tags["Repo"] = repo ? repo : "local";
    tags["Type"] = "cluster-management";
    const char* runId = std::getenv("GITHUB_RUN_ID");
    tags["RunId"] = runId ? runId : "local";
    createClusterRequest.SetTags(tags);
    
    auto createOutcome = client.CreateCluster(createClusterRequest);
    if (!createOutcome.IsSuccess()) {
        std::cerr << "Failed to create cluster in " << region << ": " 
                  << createOutcome.GetError().GetMessage() << std::endl;
        throw std::runtime_error("Unable to create cluster in " + region);
    }
    
    auto cluster = createOutcome.GetResult();
    std::cout << "Created " << cluster.GetArn() << std::endl;
    
    return cluster;
}

//#define STANDALONE_MODE
#ifdef STANDALONE_MODE
int main() {
    Aws::String region = "us-east-1";
    if (const char* env_var = std::getenv("CLUSTER_REGION")) {
        region = env_var;
    } 
    Aws::SDKOptions options;
    Aws::InitAPI(options);
    {
        try {
            auto cluster = CreateCluster(region);
            
            std::cout << "Created single region cluster:" << std::endl;
            std::cout << "Cluster ARN: " << cluster.GetArn() << std::endl;
            std::cout << "Cluster Status: " << ClusterStatusMapper::GetNameForClusterStatus(cluster.GetStatus()) << std::endl;
        }
        catch (const std::exception& e) {
            std::cerr << "Error: " << e.what() << std::endl;
        }
    }
    Aws::ShutdownAPI(options);
    return 0;
}
#endif // STANDALONE_MODE

File: python/cluster_management/src/create_single_region.py
import boto3
import os


def create_cluster(region):
    try:
        client = boto3.client("dsql", region_name=region)
        run_id = os.environ.get("GITHUB_RUN_ID", "local")
        repo = os.environ.get("GITHUB_REPOSITORY", "local")
        tags = {
            "Name": "Python-CM-Example-Single-Region",
            "Repo": repo,
            "Type": "cluster-management",
            "RunId": run_id,
        }
        cluster = client.create_cluster(tags=tags, deletionProtectionEnabled=True)
        print(f"Initiated creation of cluster: {cluster['identifier']}")

        print(f"Waiting for {cluster['arn']} to become ACTIVE")
        client.get_waiter("cluster_active").wait(
            identifier=cluster["identifier"],
            WaiterConfig={
                'Delay': 10,
                'MaxAttempts': 50
            }
        )

        return cluster
    except:
        print("Unable to create cluster")
        raise


def main():
    region = os.environ.get("CLUSTER_REGION", "us-east-1")
    response = create_cluster(region)
    print(f"Created cluster: {response['arn']}")


if __name__ == "__main__":
    main()
